### PR TITLE
Chunked encoding for snapshot status API

### PR DESCRIPTION
**Pull request by original-brownbear**
Simple implementation of chunked encoding for the snapshot status API. Tested with 100 snapshots of 25k shards (all in-progress) where it can produce the 1G+ response in less than 10s.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/SnapshotStatus.java [hunk 1]
```diff
@@ -188,24 +190,20 @@ public SnapshotStats getStats() {
...
-        builder.startObject(INDICES);
-        for (SnapshotIndexStatus indexStatus : getIndices().values()) {
-            indexStatus.toXContent(builder, params);
-        }
-        builder.endObject();
-        builder.endObject();
-        return builder;
+    public Iterator<? extends ToXContent> toXContentChunked() {
+        return Iterators.concat(Iterators.single((ToXContent) (b, p) -> {
+            var builder = b.startObject()
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/SnapshotStatus.java:195]: Nit: Isn't `b == builder`? I'd rather stick to the original style here.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/SnapshotsStatusResponse.java [hunk 2]
```diff
@@ -94,4 +87,15 @@ public boolean equals(Object o) {
...
         return snapshots != null ? snapshots.hashCode() : 0;
     }
+
+    @Override
+    public Iterator<? extends ToXContent> toXContentChunked() {
+        return Iterators.concat(
+            Iterators.single((ToXContent) (b, p) -> b.startObject().startArray("snapshots")),
+            snapshots.stream()
+                .flatMap(s -> StreamSupport.stream(Spliterators.spliteratorUnknownSize(s.toXContentChunked(), Spliterator.ORDERED), false))
+                .iterator(),
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/SnapshotsStatusResponse.java:97]: optional: this seems a pretty useful pattern, maybe we should extract a utility for it now. Kinda surprised this is the first time it appears tbh.

**Review by DaveCTurner — APPROVED** [review]
LGTM, couple of nits/suggestions

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/SnapshotsStatusResponse.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/SnapshotsStatusResponse.java:97]: We have some variations of this here and there. I have a plan for this already in the making :) PR incoming.

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**